RFC-lifecycle.md

```diff
@@ -0,0 +1,186 @@
+completely sure it's correct.
+
+## Proposed solutions
+
+Create functions that correspond to these use cases, and give them unambiguous names.
+
+### Use case 1: Resetting
+
+#### Use case
+
+Reset `nock` after a test to its initial post-`require()` state.
+
+#### Proposed solution
+
+Add `nock.reset()`.
+
+This calls:
+
+```js
+nock.restore()
+nock.cleanAll()
+nock.enableNetConnect()
+nock.activate()
+```
+
+#### Discussion
+
+`nock.reset()` is suitable for running from `afterEach()` or `finally()` to
+prepare for the next test.
+
+### Use case 2: Temporary deactivation
+
+#### Use case
+
+Temporarily disable http interception while preserving registered mocks.
+
+#### Proposed solution
+
+1. Rename `nock.restore()` to `nock.deactivate()`.
+2. Emit a deprecation warning for `nock.restore()`.
+
+#### Discussion
+
+The new name harmonizes with `nock.activate()` which is being kept: these methods are
+inverses of each other.
+
+### Use case 3: Assertions
+
+#### Use case
+
+Assert that all mocks have been satisfied.
+
+#### Proposed solution
+
+1. Add `nock.assertAllMocksUsed()`.
+2. Rename `scope.done()` to `scope.assertMocksUsed()`.
+3. Emit a deprecation warning for `scope.done()`.
+
+#### Discussion
+
+`nock.assertAllMocksUsed()` does the equivalent of `scopes.forEach(scope => scope.done())`.
+
+`scope.assertMocksUsed()` is a better name for a function that makes an
+assertion. It harmonizes better with `assertAllMocksUsed()`. Even though it's
+a method on a scope, it's `assertMocksUsed()`, not `assertMockUsed()`. This
+avoids the possibility of confusion in a case like
+`nock.get().times(5).reply()`, where a single scope is responsible for 5 mock
+requests. We need to keep this function around because it allows for granular
+control within a complex text with many scopes.
+
+`nock.assertAllMocksUsed()` is suitable to call directly from a test.
+
+Some developers may prefer to call it from an `afterEach()` hook to avoid
+boilerplate. Since not all tests in a suite may use a mock, it should do the
+natural thing when no mocks have been created, which is no-op.
+
+### Use case 4: Handling unmocked requests (proposal 1)
+
+#### Use cases
+
+- Simulate network connection failure
+- Forbid unmocked requests (Closes #844)
+
+#### Proposed solution
+
+1. Rename `nock.disableNetConnect()` to `nock.simulateUnreachability()`.
+2. Optionally pass an argument `nock.simulateUnreachability({ allowedHosts })`
+   to specify exceptions, i.e. hosts which should still be reachable.
+3. Add `nock.forbidUnmockedRequests()` which causes any unmocked request
+   to be considered a programmer error.
+4. When Nock blocks requests to mocked hostnames, throw the error instead.
+5. Emit a deprecation warning from `nock.disableNetConnect()`.
+6. Emit a deprecation warning when `nock.enableNetConnect()` is invoked with an
+   argument.
+
+#### Discussion
+
+`nock.simulateUnreachability()` emits an unreachable-like error through normal
+HTTP error channels.
```

Comment:
maybe `.simulateConnectionError()`?

Reply:
I wonder whether this should be available on a scope? So for example

```js
nock('https://google.com')
	.get('/')
	.simulateConnectionError();
```

Which would allow for some more granular control over which requests produce a network error and which don’t

Reply:
That's an interesting idea, though there would also need to be a way to configure unreachability for unmocked requests.

I think `.replyWithError()` could be used for this purpose today, and would want to think about whether it's worth adding a second way to do it.

Reply:
Yeah, I think it’s totally reasonable to also want to turn off all network connections as well. But yeah, I think it’s nice to be able to test how your application responds to one API being down/unavailable without it being blanketly based on the host.

The use case here is that I have two requests that go through an API gateway, in my case `api.ft.com`, behind that are different services which potentially could be individually unavailable. So `api.ft.com/foo` is available but `api.ft.com/bar` uses a different surface under the hood which is down. 

Currently (as far as I know) I could only simulate all of `api.ft.com` being unavailable rather than just `api.ft.com/bar`.

Reply:
Makes sense. Although, doesn't `.replyWithError()` solve that problem?

Reply:
Yeah, I didn’t actually know that existed until you’d just mentioned it. 

It does, however it might be nice to be able to not have to provide a message and have it just fail in the network and let whatever library (eg node-fetch) build the correct network error message, that it would if it was real, if that makes sense?

Reply:
Right, makes sense. It's syntactic sugar, though it keeps the dev from needing to know what kind of error is thrown when a server is unreachable.

Reply:
Yeah, exactly!